From 8e667ef87ac4bc66ab8c64599334b521655925f7 Mon Sep 17 00:00:00 2001
From: Chris Lumens <clumens@redhat.com>
Date: Thu, 14 May 2026 12:57:17 -0400
Subject: [PATCH 1/5] Med: libcrmcommon: Fix checks in localized_remote_header.

The size_total check is incorrect.  The expectation here is that it will
be equal to the size of the amount of data that was received over the
network - that is, the size of the header plus the size of the payload.
If the payload is compressed, we need that size.

Previously, there was an unwritten assumption in this check that one of
payload_compressed or payload_uncompressed would be 0, but that's not
true in the compressed case.  If the payload is compressed,
payload_uncompressed is still set because we need to know how big of a
buffer to allocate to decompress the payload.

Additionally, improve the error messages and prevent an integer overflow
when adding these sizes together before comparing.

These checks haven't failed in the past because nothing is using the
compressed payload code.
---
 lib/common/remote.c | 34 ++++++++++++++++++++++++++++++----
 1 file changed, 30 insertions(+), 4 deletions(-)

diff --git a/lib/common/remote.c b/lib/common/remote.c
index a9ce088f3f..f428c4ceda 100644
--- a/lib/common/remote.c
+++ b/lib/common/remote.c
@@ -24,7 +24,7 @@
 #include <netdb.h>
 #include <stdlib.h>
 #include <errno.h>
-#include <inttypes.h>   // PRIx32
+#include <inttypes.h>   // PRIu32, PRIx32
 
 #include <glib.h>
 #include <bzlib.h>
@@ -66,6 +66,7 @@ static struct remote_header_v0 *
 localized_remote_header(pcmk__remote_t *remote)
 {
     struct remote_header_v0 *header = NULL;
+    size_t expected_size = 0;
 
     if ((remote == NULL) || (remote->buffer == NULL)
         || (remote->buffer_offset < sizeof(struct remote_header_v0))) {
@@ -100,11 +101,36 @@ localized_remote_header(pcmk__remote_t *remote)
 
     // Sanity checks
     if (header->payload_offset != sizeof(struct remote_header_v0)) {
+        pcmk__err("Header payload offset %" PRIu32 " does not have expected "
+                  "size %zu", header->payload_offset,
+                  sizeof(struct remote_header_v0));
         return NULL;
     }
-    if ((header->payload_offset
-         + header->payload_compressed
-         + header->payload_uncompressed) != header->size_total) {
+
+    if (header->payload_compressed != 0) {
+        if (header->payload_compressed > (SIZE_MAX - header->payload_offset)) {
+            pcmk__err("Header compressed size %" PRIu32 " is too large",
+                      header->payload_compressed);
+            return NULL;
+        }
+
+        expected_size = (size_t) header->payload_offset
+                        + header->payload_compressed;
+
+    } else {
+        if (header->payload_uncompressed > (SIZE_MAX - header->payload_offset)) {
+            pcmk__err("Header uncompressed size %" PRIu32 " is too large",
+                      header->payload_uncompressed);
+            return NULL;
+        }
+
+        expected_size = (size_t) header->payload_offset
+                        + header->payload_uncompressed;
+    }
+
+    if (expected_size != header->size_total) {
+        pcmk__err("Header total size %" PRIu32 " does not match calculated "
+                  "size %zu", header->size_total, expected_size);
         return NULL;
     }
 
-- 
2.54.0

